### PR TITLE
Fix PHPStan extension for matchAllStrictGroups

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/composer/pcre/pull/34

### CODE REVIEW

File: src/PHPStan/PregMatchFlags.php [hunk 1]
```diff
@@ -34,4 +39,32 @@ static public function getType(?Arg $flagsArg, Scope $scope): ?Type
         }
         return TypeCombinator::union(...$internalFlagsTypes);
     }
+
+    static public function removeNullFromMatches(Type $matchesType): Type
```
Comment (staabm) [inline src/PHPStan/PregMatchFlags.php:43]: moved this out into the utility class to ease re-use without duplication

File: tests/PHPStanTests/UnsafeStrictGroupsCallRuleTest.php [hunk 2]
```diff
@@ -44,6 +44,10 @@ public function testRule(): void
                 'The isMatchStrictGroups call is potentially unsafe as $matches\' type could not be inferred.',
                 86,
             ],
+            [
+                'The isMatchAllStrictGroups call is unsafe as match groups "test", "1" are optional and may be null.',
```
Comment (staabm) [inline tests/PHPStanTests/UnsafeStrictGroupsCallRuleTest.php:48]: I am not 100% sure this error is expected, as I don't got the use-case of the rule

**Comment by staabm** [comment]
//cc @Seldaek

File: src/PHPStan/PregMatchTypeSpecifyingExtension.php [hunk 3]
```diff
@@ -82,19 +82,9 @@ public function specifyTypes(MethodReflection $methodReflection, StaticCall $nod
         }
 
         if (
-            in_array($methodReflection->getName(), ['matchStrictGroups', 'isMatchStrictGroups'], true)
-            && count($matchedType->getConstantArrays()) === 1
+            in_array($methodReflection->getName(), ['matchStrictGroups', 'isMatchStrictGroups', 'matchAllStrictGroups', 'isMatchAllStrictGroups'], true)
```
[the remark this answers is not in the document]
Comment (Seldaek) [reply src/PHPStan/PregMatchTypeSpecifyingExtension.php:85]: Yeah I tried that without success tho.

File: tests/PHPStanTests/UnsafeStrictGroupsCallRuleTest.php (hunk 2 above)
Comment (Seldaek) [reply tests/PHPStanTests/UnsafeStrictGroupsCallRuleTest.php:48]: This new warning is correct/fine.

Comment (Seldaek) [reply tests/PHPStanTests/UnsafeStrictGroupsCallRuleTest.php:48]: The point is, the method returns only non-null matches, but if one is null it throws, so it isn't safe to use with nullable match groups, which is what this warns about. In the test though you want nullable match groups to ensure they appear not-null in $matches.

**Comment by Seldaek** [comment]
Great thanks!

**Pull request merged**